File: Algorithm/Code_Materials/c_or_cpp/charpter10_sort/level1/quick_sort.cpp
#include <stdio.h>

void quickSort(int *array, int start, int end)
{
    if (start >= end)
    {
        return;
    }
    int left = start, right = end;
    int pivot = array[(start + end) / 2];

    while (left <= right)
    {
        while (left <= right && array[left] < pivot)
        {
            left++;
        }
        while (left <= right && array[right] > pivot)
        {
            right--;
        }
        if (left <= right)
        {
            int temp = array[left];
            array[left] = array[right];
            array[right] = temp;
            left++;
            right--;
        }
    }
    // 先处理元素再分别递归处理两侧分支，与二叉树的前序遍历非常像
    quickSort(array, start, right);
    quickSort(array, left, end);
}

int main()
{
    int array[] = {5, 2, 9, 3, 7, 6, 1, 8, 4};
    int n = sizeof(array) / sizeof(array[0]);
    quickSort(array, 0, n - 1);
    for (int i = 0; i < n; i++)
    {
        printf("%d ", array[i]);
    }
    printf("\n");
    return 0;
}

File: Algorithm/Leetcode/Code_Materials/Java/src/main/java/com/yugutou/charpter10_quicksort/QuickSort.java
package com.yugutou.charpter10_quicksort;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class QuickSort {
    public static void main(String[] args) {
        int[] array = {10,3,2,8,9,4,6,1};
//        int[] array = {8,4, 7,5,6,2,3,1};
        quickSort(array, 0, array.length - 1);
        System.out.println(Arrays.toString(array));
        List<Integer> list =new ArrayList<>();

    }

    private static void quickSort(int[] array, int start, int end) {
        if (start >= end) {
            return;
        }
        int left = start, right = end;
        int pivot = array[(start + end) / 2];
        while (left <= right) {
            while (left <= right && array[left] < pivot) {
                left++;
            }
            while (left <= right && array[right] > pivot) {
                right--;
            }
            if (left <= right) {
                int temp = array[left];
                array[left] = array[right];
                array[right] = temp;
                left++;
                right--;

            }
        }
        quickSort(array, start, right);
        quickSort(array, left, end);
    }
}
